src/apps/lwaftr/ndp.lua

```diff
@@ -0,0 +1,117 @@
+module(..., package.seeall)
+
+-- NDP address resolution.
+-- Given a remote IPv6 address, try to find out its MAC address.
+-- Exit if this fails.
+-- If resolution succeeds:
+-- All packets coming through the 'south' interface (ie, via the network card)
+-- are silently forwarded.
+-- All packets coming through the 'north' interface (the lwaftr) will have
+-- their Ethernet headers rewritten.
+
+-- Expected configuration:
+-- lwaftr <-> ipv6 fragmentation app <-> lw_eth_resolve <-> vlan tag handler
+-- That is, neither fragmentation nor vlan tagging are within the scope of this app.
+
+local constants = require("apps.lwaftr.constants")
+local lwutil = require("apps.lwaftr.lwutil")
+
+local datagram = require("lib.protocol.datagram")
+local ethernet = require("lib.protocol.ethernet")
+local ipv6 = require("lib.protocol.ipv6")
+
+
+local wr16, wr32 = lwutil.wr16, lwutil.wr32
+
+local function eth_next_is_ipv6(pkt)
+   return rd16(pkt.data + 12) == constants.n_ethertype_ipv6
+end
+
+local function ipv6_next_is_icmp6(pkt)
+   local offset = 19 -- 14 Ethernet + 6 IPv6 header bytes, 0-indexed
+   return pkt[offset] == constants.proto_icmpv6
+end
+
+local function icmpv6_type_is_na(pkt)
+   local offset = 14 + 40 - 1
+   return pkt[offset] == constants.icmpv6_na
+end
+
+-- The relevant byte is:
+-- [ R S O 0 0 0 0 0 0 ], where S = solicited
+local function is_solicited_na(pkt)
+   local offset = 14 + 40 + 3
+   return bit.band(0x40, pkt[offset]) == 1
+end
+
+-- TODO: tune this for speed
+-- Note that this must be run after any vlan tags have been stripped out
+-- and after any reassembly has been done, and no extra IPv6 headers
+-- must be present
+--- TODO: could this reasonably use pflang?
+function is_solicited_neighbor_announcement(pkt)
+   return pkt.length > 57 and
+          eth_next_is_ipv6(pkt) and
+          ipv6_next_is_icmp6(pkt) and
+          icmpv6_type_is_na(pkt) and
+          is_solicited_na(pkt)
+end
+
+function get_dst_ethernet(p, target_ipv6_addr)
+   local na_addr_offset = 14 + 40 + 7
+   if C.memcmp(target_ipv6_addr, pkt.data + na_addr_offset) == 0 then
```

Comment:
C.memcmp is slower than an inline ipv6_equals because it forces register saves and is an optimization barrier. I guess we need to merge in https://github.com/Igalia/snabbswitch/pull/198.